src/common/protobuf_utils.hpp

```diff
@@ -64,6 +64,53 @@ struct Slave;
 
 namespace protobuf {
 
+//
+// A "protobuf union" is a protobuf message
+// which has an enum named "Type" defined in its scope,
+// a required field named "type" of a "Type" type
+// and a field corresponding to each member of the "Type" enum.
+//
+// A valid "protobuf union" should satisfy the following requirements.
+// 1. A member of the Type enum with a number (not index!) of 0,
+//    if there is any, should be named "UNKNOWN".
+// 2. For each other member of the Type enum there should be an optional field
+//    in the message with an exactly matching name in lowercase.
+// 3. The fields that correspond to members of the Type enum other than
+//    message.type(), should not be set.
+//
+// NOTE: If possible, oneof should be used in the new messages instead
+// of the "protobuf union".
+//
+class UnionValidator
+{
+public:
+  UnionValidator(
+      const google::protobuf::Descriptor*,
+      const google::protobuf::EnumDescriptor*);
+
+  Option<Error> validate(
+      const int messageTypeNumber, const google::protobuf::Message&) const;
+
+private:
+  std::vector<std::pair<int, const google::protobuf::FieldDescriptor*>>
+    unionFieldDescriptors_;
+  const google::protobuf::EnumDescriptor* typeDescriptor_;
+};
+
+
+// This function returns Error if the protobuf union is not valid
+// and Nothing otherwise.
+// In case MessageWithUnion is not a protobuf union,
+// this function will abort the process on the first use.
+template <class MessageWithUnion>
+Option<Error> validateProtobufUnion(const MessageWithUnion& message)
```

Comment:
This might be easy to miss from reading the implementation:
```
// NOTE: A protobuf union is only invalid when unexpected member fields are set.
// It is valid when all members of a protobuf union are empty.
```

Reply:
I rewrote the comment which explains what a "protobuf union" is almost completely. I hope now this point is more difficult to miss.